[Button-367] Refactoring

robergro opened the pull request:
Content:
- Update ViewModels
- Update View and add IconButton
- Add SwiftUI version
- Update snapshot testing
- Update demo

This pull request has:
- [x] Code documentation on publics
- [x] Accessibility
	- [x] Dynamic sizes
	- [x] Identifiers
- [x] Tests
- [x] Demo update
- [ ] Wiki

michael-zimmermann commented on core/Sources/Components/Button/Manager/Main/ButtonMainManagerTests.swift:
```diff
@@ -0,0 +1,174 @@
+
+// MARK: - Stub
+
+private struct Stub {
```
This doesn't stub anything. Its more of a SutFactory,

robergro replied:
Check my global message for this PR :)

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Internal/ButtonContentView.swift:
```diff
@@ -0,0 +1,122 @@
+            colorToken: self.viewModel.currentColors?.borderColor ?? ColorTokenDefault.clear
+        )
+        .compositingGroup()
+        .disabled(!(self.viewModel.state?.isUserInteractionEnabled ?? true))
```
.disable(self.viewModel.state?.isUserInteractionEnabled == false)

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Internal/ButtonContentView.swift:
```diff
@@ -0,0 +1,122 @@
+
+    private let manager: ButtonMainManager<ViewModel>
+
+    private var viewModel: ButtonMainViewModel {
```
I would have expected the viewModel to be an observable object, since it appears, that the published values it contains change

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Public/Button/ButtonView.swift:
```diff
@@ -0,0 +1,199 @@
+
+    // MARK: - Private Properties
+
+    @ObservedObject private var manager: ButtonManager
```
I don't think the manager needs to be an ObservedObject

robergro replied:
Check my global message for this PR :)

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Public/Button/ButtonView.swift:
```diff
@@ -0,0 +1,199 @@
+
+    @ViewBuilder
+    private func title() -> some View {
+        if let text = self.manager.controlStateText.text {
```
All these states are quite confusing.
The current struct has a reference to the controlStateText which is an observed object. The manager doesn't have a published object controlStateText. I think this may be triggered because of the current observed object controlStateText, but it appears to be accidental.

robergro replied:
Check my global message for this PR :)

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Public/Button/ButtonView.swift:
```diff
@@ -0,0 +1,199 @@
+
+    @ObservedObject private var manager: ButtonManager
+
+    @ObservedObject private var viewModel: ButtonViewModel
```
I find it difficult to find a clear responsibility of all the observed objects. I would try and get rid of the manager and just have a viewModel which delegates to the control states, and keep this oblivious to the view. In my opinion, the view should only interact with the view model and not four objects, which all seem to be interconnected.

robergro replied:
Check my global message for this PR :)

michael-zimmermann commented on core/Sources/Components/Button/View/SwiftUI/Internal/ButtonContentView.swift:
```diff
@@ -0,0 +1,122 @@
+import SwiftUI
+import Foundation
+
+struct ButtonContentView<ContentView: View, ViewModel: ButtonMainViewModel>: View {
```
Maybe this is more of a `ButtonContainerView` since the content is injected.

michael-zimmermann reviewed (commented):
The state changes in the demo don't work for the icon button.

E.g, when I press the button, no image is shown.
<img width="343" alt="Bildschirmfoto 2023-12-28 um 12 40 03" src="https://github.com/adevinta/spark-ios/assets/128726463/88ac177c-cad9-4786-80fb-69bb53b9676b">

The same problems exist with the disabled and selected states.

michael-zimmermann reviewed (commented):
In the demo the UIKit button doen't change it's appearance whether it's in selected state or not.
<img width="381" alt="Bildschirmfoto 2023-12-28 um 12 45 01" src="https://github.com/adevinta/spark-ios/assets/128726463/abab049e-a9b6-458c-b225-754cadc604f2">

michael-zimmermann reviewed (commented):
In the SwiftUI button there are some state changes not reflected:
Given:
- normal state: none
- highlighted state: text/Image ....
When: the button is pressed
Then: the button doesn't change
Given: 
- normal state: Image and attributed text
- highlighted: none
When: the button is pressed
Then: an icon is shown

And different combinations such as above

robergro commented on core/Sources/Components/Button/Manager/Main/ButtonMainManager.swift:
```diff
@@ -0,0 +1,68 @@
+
+    let viewModel: ViewModel
+
+    private(set) var control = Control()
```
OoO you're right, I created a ControlStatus, I will reuse this object and remove the `Control`.

On the ControlStatus, the init contains the isEnabled because I wanted to reuse the same variable names as for the UIControl (isEnabled). 
But, because it creates confusion, I replace isDisabled to isEnabled in property

robergro commented on core/Sources/Components/Button/Manager/Main/ButtonMainManager.swift:
```diff
@@ -0,0 +1,68 @@
+import SwiftUI
+
+/// Manage the SwiftUI Button View.
+class ButtonMainManager<ViewModel: ButtonMainViewModel>: ObservableObject {
```
Check my global message for this PR :)

robergro commented:
@michael-zimmermann,
Thank you for all your comments 💪 

So, I update the code:
- I removed the manager, I replaced it with viewModel protocol (ButtonMainSUIViewModel) and two viewModel (IconButtonSUIViewModel & ButtonSUIViewModel) only for SwiftUI.
- There is no more multiple observable object on the views (just the viewmodel)
- On the demo app: it is no this is no longer possible to have no text in the **button** for the **normal state**.
- On the demo app: it is no this is no longer possible to have no image in the **icon button** for the ** for the normal state**.

michael-zimmermann commented on core/Sources/Components/Button/ViewModel/Button/ButtonSUIViewModel.swift:
```diff
@@ -0,0 +1,42 @@
+
+    // MARK: - Properties
+
+    var controlStatus: ControlStatus = .init()
```
`var controlStatus = ControlStatus()` ?

michael-zimmermann reviewed (approved):
Forget the small comment I left.
👍  Looks good

robergro commented on core/Sources/Common/Control/Status/ControlStatus.swift:
```diff
@@ -7,22 +7,34 @@
 
 /// The current status of the control: highlighted or not, disabled or not and selected or not.
-struct ControlStatus {
+final class ControlStatus: Equatable {
```
I need to use **class** to have the possibility to update the object on a class extension (here the ButtonMainSUIViewModel). It is not possible with a struct, because the object is immutable
